Fix #2455: Add lockfile functionality to prevent user error

dozure opened the pull request:
`scripts/start.sh` and `scripts/install_prerequisites.sh` now use a lockfile mechanism to keep users from breaking the install under vagrant
Replaced a `return 1` outside of a function call with an `exit 1`
Added new `.lock` file to `.gitignore`

seanlip commented:
Hi @bbriggs, PTAL? Thanks!

dozure commented:
Not sure what to do here. The travis job that failed looks like a bad test in a completely unrelated bit of code.

seanlip commented:
Ah, Travis is sometimes flaky. I've restarted it.

BenHenning commented:
Thanks @dozure, this looks good to fix the vagrant issue but has a potential unintended side effect. With these changes, running ``start.sh`` twice will result in the lock error message. Is this WAI? There's a legitimate use case where ``start.sh`` is run while a dev server is currently running (such as updating dependencies when switching branches). Can you factor this use case into the changes you made?

dozure commented:
Thanks for the feedback, @BenHenning!

Ah - I didn't know that, it didn't come up un discussion. Only real way I can think of to do that offhand, is adding to my warning a message asking the user to pass a flag (say -l or something) to override. Probably only be be annoying the first time someone does it on purpose then they'll remember.

Since this is only really a problem under vagrant, I could also wrap the whole lockfile thing in a `if $is_vagrant` (or whatever the magic variable is that vagrant sets) check to avoid annoying people working on local machines/unit tests.

Either or both of those sound ok?

seanlip commented:
I like the "if is_vagrant" approach better; I think we shouldn't worsen the dev experience for non-vagrant developers.

dozure commented:
Ok, sounds good. Is creating a new file ok? Since this is happening in `install_prerequisites.sh` and `setup.sh` it would be cleaner to have another file, say `vagrant_lock.sh` that I source if I detect vagrant, but I don't know what the policy on adding new files, filesystem clutter, naming conventions, etc are.

seanlip commented:
I think adding a new file is fine, yep!

seanlip commented:
Oh, btw -- should the change be to setup.sh instead of start.sh? I think most of our other scripts source setup.sh but not all of them use start.sh; this may need a bit of analysis.

dozure commented:
When vagrant boots, it really only does two things, run `scripts/install_prerequisites.sh` and `scripts/start.sh`; only `start.sh` sources `setup.sh`. We need the check in both since depending on how long it takes the user to open a new shell and type `vagrant ssh` and then (erroneously) figure out that they need to run `start.sh` either of the two scripts could be running.

I could put the vagrant check into `setup.sh` but then I'd need to have `install_prerequisites` also source `setup.sh`, though given all that that script does, it seems heavy-handed to run it twice.

seanlip commented:
Ah, I see. How about having a separate script whose only purpose is to handle the locking on Vagrant, and sourcing that file from multiple places? If it is possible to do that, that might be cleaner than duplicating the functionality in multiple files.

dozure commented:
Yep, that's what I meant in https://github.com/oppia/oppia/pull/2749#issuecomment-262661974. Sorry if I wasn't clear, any my typo of `setup.sh` probably only added to confusion. Working on migrating it to a new file and only sourcing it if vagrant is detected.

seanlip commented:
OK, sounds good. Thanks!

dozure commented:
@BenHenning @seanlip PTAL? Moved the code to another file to be sourced, wrapped the whole thing in a check for vagrant, and added a --nolock parameter just in case a dev working under vagrant wants to run it twice on purpose.

I copied/pasted the license boilerplate from another bash file. I updated the year to 2016; not sure if that was the right thing to do or not.

seanlip commented:
Passing to @BenHenning for review.

BenHenning commented on scripts/vagrant_lock.sh:
```diff
@@ -0,0 +1,65 @@
+# It can be overridden by passing --nolock to start.sh
+
+
+# This keeps all the arguments passed in, except for --nolock just in case
```
Stripping the nolock seems a bit complicated. Could we maybe have a flag less likely to have problems downstream, like ``novagrantlock`` to avoid the extra complexity? Bash can get complicated quickly, which makes it harder to maintain in the future.

BenHenning replied:
@bbriggs do you see any reason why we would care about the number of arguments or order? If not, I think we should just remove the nolock stripping here, since it is a bit complicated.

BenHenning replied:
Let's remove the noblock stripping. Thanks @dozure.

BenHenning commented on scripts/vagrant_lock.sh:
```diff
@@ -0,0 +1,65 @@
+
+set -- "${keep_args[@]}"
+
+[ $NO_LOCK ] && return 0
```
Prefer using an ``if`` block since it's a bit more readable, and tends to be waht we do in these situations.

dozure replied:
Done. Holding commit until I can push both.

dozure commented:
Sorry for the radio silence recently. I had some unexpected stuff come up in real life, plus the holidays here in the US. I've removed the nolock stripping and changed the test && bashism to an if-then.
